fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/ElementShader.java

```diff
@@ -0,0 +1,105 @@
+package net.minecraftforge.fml.earlydisplay;
+
+
+import static org.lwjgl.opengl.GL32C.*;
+
+public class ElementShader {
+    private int program;
+    private int textureUniform;
+    private int screenSizeUniform;
+    private int renderTypeUniform;
+
+    public void init() {
+        int vertexShader = glCreateShader(GL_VERTEX_SHADER);
+        int fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+
+        // Bind the source of our shaders to the ones created above
+        glShaderSource(fragmentShader, """
+                 #version 330 core
```

Comment:
What does that mean? This is my first GL code.

Reply:
You are using OpenGL 3.3, or OpenGL 3.2 + GL_ARB_explicit_attrib_location features in these shaders.
To be OpenGL 3.2 compliant this would need to use `#version 150` and not make any use of `layout(location= xyz)` instead being bound before linking the program via glBindAttribLocation and glBindFragDataLocation.

This is true for all shaders/programs.